### PR TITLE
feat(user): add support for ssl_certificate and other auth types

### LINKED ISSUE #136: Allow users to authorize via cert
Basically the equivalent of `users.xml` configuration that looks like this:
```
<miniaczq>
    <ssl_certificates>
        <common_name>285746391827364246783</common_name>
    </ssl_certificates>
</miniaczq>
```

Few useful cases:
- Service agents and users temporary access
- Users using a single cert for access to multiple services

### PR DESCRIPTION
**Pull request by sepandr**
## Summary

Replaces the single `password_sha256_hash` capability with a nested `auth {}` block that supports multiple, combinable authentication methods.

- Per-method sub-blocks: `no_password`, `plaintext_password`, `sha256_password`, `sha256_hash`, `double_sha1_password`, `double_sha1_hash`, `bcrypt_password`, `bcrypt_hash`, `ssl_certificate` (`common_name` or `subject_alt_name`), `http` (`server` or `scheme`), `ssh_key` (`public_key` + `type`), `ldap` (`server`), and `kerberos` (optional `realm`).
- Every method except `no_password` may be repeated; all configured methods render into a single `IDENTIFIED WITH m1, m2, …` clause. `no_password` is exclusive.
- Secret methods take either `value` (stored in state; Terraform/OpenTofu < 1.11) or write-only `value_wo` + `value_wo_version` (>= 1.11); bump the version to re-apply. `*_hash` methods validate their value format.
- **Secrets never appear in query text or logs**: password/hash values are sent as server-side query parameters (`IDENTIFIED WITH sha256_hash BY {secret_0:String}`) — bound via `clickhouse-go` `WithParameters` on the native protocol and `param_*` URL parameters on HTTP. HTTP transport errors have the URL query stripped so parameter values cannot leak through error messages.
- Auth changes apply in place via `ALTER USER` — no resource replacement.
- Legacy `password_sha256_hash` / `password_sha256_hash_wo` (+ `_version`) remain, are deprecated, and compose additively as an extra `sha256_hash` method, so existing configurations keep working.
- Validation is fully declarative (schema-level). Secret values can't be read back from ClickHouse, so external drift is not detected; on import only the user identity is read and methods are re-asserted from config on next apply.

Closes #136

## Changes

- `internal/querybuilder/authmethod.go` (+ `authmethod_test.go`) — method render-spec table and `identifiedClause()` for multi-method `IDENTIFIED WITH`; secret args render as `{secret_N:String}` placeholders and are returned as a parameters map.
- `internal/querybuilder/createuser.go` / `alteruser.go` — accept `[]AuthMethod`, render `IDENTIFIED` for create and in-place alter, expose `Parameters()`.
- `internal/clickhouseclient` — `Exec` accepts optional query parameters; native binds them server-side via `clickhouse.WithParameters`, HTTP sends them as `param_*` URL parameters; transport-error URLs are redacted.
- `internal/dbops/user.go` — resolve and (re)assert auth methods on create/update, passing secret parameters through to the client.
- `pkg/resource/user/auth.go` — `auth` schema blocks, resource config validators, and plan/config → `dbops.AuthMethod` resolution.
- `pkg/resource/user/model.go`, `pkg/resource/user/user.go` — model types, schema wiring, in-place update.
- `internal/testutils/resourcebuilder` — nested-block and multi-arg function support for tests.
- `pkg/resource/user/user_test.go` — acceptance and declarative-validation tests.
- docs and examples updated for the `auth` block.

## Test plan

- [x] Unit tests (`go test ./...`), incl. `identifiedClause` placeholder/parameter pairing, builder `Parameters()`, and a regression test that HTTP transport errors don't leak parameter values
- [x] Querybuilder tests cover all methods and `IDENTIFIED WITH` rendering edge cases
- [x] Acceptance: create via native/HTTP on single/replicated/localfile; `sha256_hash` write-only; `ssl_certificate`; multiple methods; `no_password`; in-place `sha256_hash` rotation and `ssl_certificate` change
- [x] Validation: `no_password` exclusivity, `value`/`value_wo` exactly-one, `ssl_certificate` CN/SAN exactly-one, hash-format validation, at-least-one-method

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by sepandr** [comment]
@whites11 @nunoadrego @gkorolev requesting reviews please 🙏

**Comment by nunoadrego** [comment]
> @whites11 @nunoadrego @gkorolev requesting reviews please 🙏

We will take a look. Have you had a chance to test these changes locally? If so, could you share the setup so we can double-check it more quickly?

**Comment by sepandr** [comment]
@nunoadrego yes, I've tested this end-to-end against a local ClickHouse instance. Sharing the full setup below so you can reproduce.

## Setup

**1. Start ClickHouse locally with `access_management` enabled** (needed so the `default` user can `CREATE USER`):

`users-override.xml`:
```xml
<?xml version="1.0"?>
<clickhouse>
  <users>
    <default>
      <password>testpassword</password>
      <access_management>1</access_management>
      <named_collection_control>1</named_collection_control>
      <show_named_collections>1</show_named_collections>
    </default>
  </users>
</clickhouse>
```

```bash
docker run -d --name clickhouse-test \
  -p 18123:8123 -p 19000:9000 \
  -v $(pwd)/users-override.xml:/etc/clickhouse-server/users.d/access.xml \
  --ulimit nofile=262144:262144 \
  clickhouse/clickhouse-server:latest
```

**2. Build the provider and point Terraform at it:**

```bash
CGO_ENABLED=0 go build -o /tmp/prov/terraform-provider-clickhousedbops
```

`dev.tfrc`:
```hcl
provider_installation {
  dev_overrides {
    "ClickHouse/clickhousedbops" = "/tmp/prov"
  }
  direct {}
}
```

**3. Apply this config** (`main.tf`):

```hcl
provider "clickhousedbops" {
  protocol = "native"
  host     = "localhost"
  port     = 19000
  auth_config = {
    strategy = "password"
    username = "default"
    password = "testpassword"
  }
}

resource "clickhousedbops_user" "legacy_hash_user" {
  name                 = "legacy_hash_user"
  password_sha256_hash = sha256("legacy_password")
}

resource "clickhousedbops_user" "cert_user" {
  name       = "cert_user"
  auth_type  = "ssl_certificate"
  auth_value = "CN=test-service,OU=Engineering"
}

resource "clickhousedbops_user" "plaintext_user" {
  name       = "plaintext_user"
  auth_type  = "plaintext_password"
  auth_value = "plaintext-secret-value"
}

resource "clickhousedbops_user" "auth_type_sha256_user" {
  name       = "auth_type_sha256_user"
  auth_type  = "sha256_hash"
  auth_value = sha256("another_password")
}

resource "clickhousedbops_user" "no_password_user" {
  name      = "no_password_user"
  auth_type = "no_password"
}
```

```bash
TF_CLI_CONFIG_FILE=./dev.tfrc terraform apply -auto-approve
```

## Verification

After `apply`, querying `system.users`:

```
┌─name──────────────────┬─auth_type──────────────┬─auth_params─────────────────────────────────────────────┐
│ plaintext_user        │ ['plaintext_password'] │ ['{}']                                                  │
│ cert_user             │ ['ssl_certificate']    │ ['{"common_names":["CN=test-service,OU=Engineering"]}'] │
│ legacy_hash_user      │ ['sha256_password']    │ ['{}']                                                  │
│ no_password_user      │ ['no_password']        │ ['{}']                                                  │
│ auth_type_sha256_user │ ['sha256_password']    │ ['{}']                                                  │
└───────────────────────┴────────────────────────┴─────────────────────────────────────────────────────────┘
```

And `SHOW CREATE USER` confirms the SQL is correct — note the `CN` syntax for `ssl_certificate`:

```
CREATE USER cert_user IDENTIFIED WITH ssl_certificate CN 'CN=test-service,OU=Engineering'
CREATE USER plaintext_user IDENTIFIED WITH plaintext_password
CREATE USER auth_type_sha256_user IDENTIFIED WITH sha256_password
CREATE USER no_password_user IDENTIFIED WITH no_password
CREATE USER legacy_hash_user IDENTIFIED WITH sha256_password
```

Also confirmed:
- `terraform destroy` cleanly removes all 5 users.
- Validation errors fire as expected when mixing `auth_type` with `password_sha256_hash`, and when passing an invalid `auth_type` (e.g. `"not_a_real_type"`).
- Backward compatibility preserved — the legacy `password_sha256_hash` path still produces the same `sha256_password` user as before.

Let me know if you'd like me to run any additional scenarios.

**Comment by sepandr** [comment]
@whites11 @nunoadrego @gkorolev bump on reviews please 🙏

File: pkg/resource/user/user.go [hunk 1]
```diff
@@ -99,6 +99,35 @@ func (r *Resource) Schema(_ context.Context, _ resource.SchemaRequest, resp *res
...
+						path.MatchRoot("password_sha256_hash_wo"),
+					),
+				},
+			},
+			"auth_value": schema.StringAttribute{
+				Optional:    true,
+				Sensitive:   true,
+				Description: "Authentication value for the user. The meaning depends on auth_type: for sha256_hash it's the hash, for ssl_certificate it's the CN (Common Name), for plaintext_password it's the password, etc. Not required when auth_type is no_password.",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
```
Comment (nunoadrego) [inline pkg/resource/user/user.go]: Maybe we can add:
```
Validators: []validator.String{
        stringvalidator.AlsoRequires(path.MatchRoot("auth_type")),
    },
```

File: docs/resources/user.md [hunk 2]
```diff
@@ -52,6 +71,19 @@ resource "clickhousedbops_user" "john" {
   # You'll want to generate the password and feed it here instead of hardcoding.
   password_sha256_hash = sha256("test")
 }
+
+# Example using ssl_certificate authentication (e.g., for Teleport mTLS)
```
Comment (nunoadrego) [inline docs/resources/user.md]: Let's remove teleport references. Use generic name for examples.

File: docs/resources/user.md [hunk 3]
```diff
@@ -65,6 +97,8 @@ resource "clickhousedbops_user" "john" {
 
 > **NOTE**: [Write-only arguments](https://developer.hashicorp.com/terraform/language/resources/ephemeral#write-only-arguments) are supported in Terraform 1.11 and later.
 
+- `auth_type` (String) Authentication type for the user. Supported values: sha256_hash, ssl_certificate, plaintext_password, bcrypt_hash, double_sha1_hash, no_password. When set, auth_value must also be provided (except for no_password). Conflicts with password_sha256_hash and password_sha256_hash_wo.
+- `auth_value` (String, Sensitive) Authentication value for the user. The meaning depends on auth_type: for sha256_hash it's the hash, for ssl_certificate it's the CN (Common Name), for plaintext_password it's the password, etc. Not required when auth_type is no_password.
```
Comment (nunoadrego) [inline docs/resources/user.md]: How can we also support write-only mode? Otherwise, the new `auth_value` field will always be written to the state.

File: pkg/resource/user/user.go [hunk 4]
```diff
@@ -176,24 +205,36 @@ func (r *Resource) Create(ctx context.Context, req resource.CreateRequest, resp
...
+	}
+
 	switch {
+	case !plan.AuthType.IsNull():
+		// New auth_type/auth_value fields
+		user.AuthType = plan.AuthType.ValueString()
+		if !plan.AuthValue.IsNull() {
+			user.AuthValue = plan.AuthValue.ValueString()
+		}
+		if user.AuthType != "no_password" && user.AuthValue == "" {
```
Comment (nunoadrego) [inline pkg/resource/user/user.go]: `no_password` + non-null `auth_value` silently accepted:
```
if user.AuthType == "no_password" && user.AuthValue != "" {
    resp.Diagnostics.AddError(
        "Invalid auth_value",
        "auth_value must not be specified when auth_type is no_password",
    )
    return
}
```

**Review by nunoadrego — COMMENTED** [review]
Thanks for opening this PR. Left some comments for consideration.

**Comment by tmateen** [comment]
Hi I see this PR is stale but it’d be a great feature to have - specifically, if it would be possible to build on this & also add in [ssh auth](https://clickhouse.com/docs/knowledgebase/how-to-connect-to-ch-cloud-using-ssh-keys) for users? Let me know if it would make sense to add to this PR/ open a new one incase we would not like to go this route..

File: pkg/resource/user/user.go (hunk 1 above)
Comment (sepandr) [reply pkg/resource/user/user.go]: Done in 5b3d6ee — `auth_value` now has `stringvalidator.AlsoRequires(path.MatchRoot("auth_type"))`, plus a `ConflictsWith` against the new `auth_value_wo`. Added a plan-only validation test covering `auth_value` without `auth_type`.

File: docs/resources/user.md (hunk 2 above)
Comment (sepandr) [reply docs/resources/user.md]: Done in 5b3d6ee — removed all Teleport references from the docs, examples, and test fixtures; examples now use generic names (`cert_user`, `cert-common-name`).

File: docs/resources/user.md (hunk 3 above)
Comment (sepandr) [reply docs/resources/user.md]: Done in 5b3d6ee — added `auth_value_wo` (`WriteOnly: true`) + `auth_value_wo_version`, mirroring the existing `password_sha256_hash_wo`/`password_sha256_hash_wo_version` pattern. `Create` reads the write-only value from the config, and it conflicts with `auth_value` (which is kept for Terraform/OpenTofu < 1.11, same split as the password fields). Docs regenerated and validation + acceptance tests added.

Verified end-to-end against a local ClickHouse: `auth_value_wo = "CN=..."` creates the user with the right `common_names` in `system.users`, and the value does not appear anywhere in `terraform.tfstate` (only `auth_value_wo_version` is stored).

File: pkg/resource/user/user.go (hunk 4 above)
Comment (sepandr) [reply pkg/resource/user/user.go]: Done in 5b3d6ee — added exactly this check in `Create`, and also a `ValidateConfig` implementation so the `no_password` + `auth_value`/`auth_value_wo` combination is rejected at plan time when the value is known (the `Create` check still covers values unknown during plan). Both paths have plan-only validation tests.

**Comment by sepandr** [comment]
@nunoadrego thanks for the review — all comments addressed in 5b3d6ee:

- `auth_value` now requires `auth_type` (`AlsoRequires` validator)
- Teleport references removed from docs/examples/test fixtures
- Added write-only `auth_value_wo` + `auth_value_wo_version` following the `password_sha256_hash_wo` pattern, so the auth value can be kept out of state (`auth_value` remains for Terraform/OpenTofu < 1.11)
- `no_password` + `auth_value`/`auth_value_wo` is now rejected, both at plan time via `ValidateConfig` and in `Create` for values unknown during plan

Re-verified end-to-end against a local ClickHouse using the setup posted above, plus the new write-only path: the CN lands in `system.users` and the write-only value is absent from `terraform.tfstate`. Unit tests and the plan-only validation acceptance tests pass locally on Terraform 1.11.3.

@tmateen ssh_key auth needs a different SQL shape (`BY KEY '<key>' TYPE '<type>'`, potentially multiple keys), so it doesn't fit the single `auth_value` field cleanly — I'd suggest a follow-up PR building on the `Identification` plumbing this one adds.

**Comment by tmateen** [comment]
@sepandr Thanks for the response & insights - sounds good, will look to add a separate PR for that.

On another note, I’ve realised that Clickhouse has the capability to allow multiple auth mechanisms per user; i.e., one could use something like 

```
CREATE USER IF NOT EXISTS <USER> identified with ssh_key BY KEY '<PUBLIC_KEY_HERE>' TYPE ‘<TYPE>', bcrypt_password by '<PASSWORD>’ , ...;
```

In order to be able to support that with the clickhouse_dbops provider, would it make sense to have a different input field per authentication type instead?

**Comment by sepandr** [comment]
> @sepandr Thanks for the response & insights - sounds good, will look to add a separate PR for that.
> 
> On another note, I’ve realised that Clickhouse has the capability to allow multiple auth mechanisms per user; i.e., one could use something like
> 
> ```
> CREATE USER IF NOT EXISTS <USER> identified with ssh_key BY KEY '<PUBLIC_KEY_HERE>' TYPE ‘<TYPE>', bcrypt_password by '<PASSWORD>’ , ...;
> ```
> 
> In order to be able to support that with the clickhouse_dbops provider, would it make sense to have a different input field per authentication type instead?

I didn't realize that was a feature! Yeah maybe this could get converted into accepting a list per user, and then the provider looking for the appropriate input_field given the `auth_type` @tmateen

**Comment by GrigoryPervakov** [comment]
I've patched this branch, so it should be ready for merge now. But then I started to think about supporting multiple auth methods, and it seems that it is better to support it right here to not introduce attributes that would be obsolete shortly.
I have a few ideas; which looks better?
```hcl
# multiple nested auth blocks with the same schema
# Simple implementation for now, but harder to enforce validation, and if we'd decide to support methods with more than one arg would look bulky.
# For methods which arg is not a secret value, it would still be redacted in plan
resource "clickhousedbops_user" "auth_type_sha256_user" {
  name = "auth_type_sha256_user"
  auth  { 
    type  = "sha256_hash"
    value = sha256("another_password")
  }
  auth  { 
    type  = "plaintext_password"
    value_wo = sha256("another_password")
    value_version = 1
  }
}

# per-type nested blocks with dedicated schema
# more code, but allows validation of each field and masking only needed attributes
resource "clickhousedbops_user" "auth_type_sha256_user" {
  name = "auth_type_sha256_user"
  sha256_hash  { 
    value = sha256("another_password")
  }
  ssl_certificate {
    common_name = "a"  { 
  }
  ssl_certificate {
    common_name = "b"  { 
  }
  ssh_key {
    public_key = "...",
    type = "ssh-rsa"
  }
}

# same but in the separate auth block
resource "clickhousedbops_user" "auth_type_sha256_user" {
  name = "auth_type_sha256_user"
  auth {
    sha256_hash  { 
      value = sha256("another_password")
    }
    ssl_certificate {
      common_name = "a"  { 
    }
    ssl_certificate {
      common_name = "b"  { 
    }
    ssh_key {
      public_key = "...",
      type = "ssh-rsa"
    }
  }
}
```

**Comment by tmateen** [comment]
@GrigoryPervakov Thanks for following up! I think the last option (separate auth block + sub-blocks for different auth types) may be the cleanest way to support multiple auth types with different fields (i.e., also supporting the ssh_key value).

**Review by nunoadrego — APPROVED** [review]

**Comment by sepandr** [comment]
Thank you @GrigoryPervakov for taking this over 😄
